File: app/lib/model/detection.dart
import 'package:app/model/prediction.dart';
import 'package:uuid/uuid.dart';

class Detection {
  late String _id;
  late List<Prediction> _predictions;
  late String _filePath;
  late DateTime _start;
  late DateTime _end;

  Detection(this._predictions, this._filePath, {String? id}) {
    _id = id ?? const Uuid().v4();
    _start = _predictions[0].receivedAt;
    _end = _predictions[_predictions.length - 1].receivedAt;
  }

  Detection.fromMap(Map<String, dynamic> map) {
    _id = map["uid"];
    _filePath = map["filePath"];
    _predictions = [];
    for (int i = 0; i < map["predictions"].length; i++) {
      Prediction prediction = Prediction(
        map["predictions"][i]["filePath"]!,
        double.parse(map["predictions"][i]["value"]!),
        id: map["predictions"][i]["id"],
        receivedAt: DateTime.parse(map["predictions"][i]["receivedAt"]!),
      );
      _predictions.add(prediction);
    }
    _start = _predictions[0].receivedAt;
    _end = _predictions[_predictions.length - 1].receivedAt;
  }

  String get id => _id;
  List<Prediction> get predictions => _predictions;
  String get filePath => _filePath;
  DateTime get start => _start;
  DateTime get end => _end;
}

File: web_app/api/model/detection.py
import time
import datetime
import uuid
from .prediction import Prediction
class Detection:
    uid:str
    predictions:list[Prediction]
    filePath:str
    start:time
    end:time

    def __init__(self, predictions:[],filePath:str="",uid:str=None, start:time=None, end:time=None):
        if uid == None:
            self.uid = str(uuid.uuid4())
        else:
            self.uid = uid
        self.predictions = predictions
        self.filePath = filePath
        if start == None:
            self.start = self.predictions[0].receivedAt
        if end == None:
            self.end = self.predictions[len(predictions)-1].receivedAt

    def toDict(self):
        predictionList = []
        for pred in self.predictions:
            predictionList.append(pred.toDict())
        return {
            "uid":self.uid,
            "predictions": predictionList,
            "filePath":self.filePath,
            "start":str(datetime.datetime.fromtimestamp(self.start)),
            "end":str(datetime.datetime.fromtimestamp(self.end))
        }